### PR TITLE
Peer delete did not trigger ZEBRA_NEXTHOP_UNREGISTER to ZEBRA results…

### PR DESCRIPTION
**Pull request by vishaldhingra**
… in loss of route node objects in zebra

RCA : bgp_nexthop_cache is allocated and ZEBRA_NEXTHOP_REGISTER is sent to ZEBRA.

The association of BGP is as below :
(nexthop_cache_table)<---->bgp_nexthop_cache---->peer
                                |<--------------> bgp_info

for peer association bnc has nht_info
for bgp_info bnc has list count

Allocation of bnc :
Scenario 1: When peer pointer exists
Scenario 2: When peer pointer does not exists and bgp_info exists via bgp_update in IBGP session, when next-hop is
            not self.

Allocation is fine in all scenarios

Deletion of bnc:
Scenario 1: When peer pointer exists
Scenario 2: When peer pointer does not exists

The deletion is triggered via bgp_unlink_nexthop_by_peer()
In this context bgp->nht_info sets to NULL, which is correct as peer is going to delete.
but in bgp_unlink_nexthop_check() expects the nht_info should not be NULL.
So it will not call bgp_unlink_nexthop_check to ZEBRA

Fix :

When bgp_unlink_nexthop_check() gets called;
>> Then it should aware that PEER was there or not.
If PEER was there then it has to wait let the peer delete and list count of bgp_info go to zero
>>> If PEER was not there
then it should not check PEER and rely only on bgp_info list count.

In order to do this, I have introduced two flags to know the PEER was there or not.

The code checked-in in this context was fine to avoid the crash, but ZEBRA_NEXTHOP_UNREGISTER is missed out.

This fix will also take care of this scenario, i.e it will take care work queue handling of
bgp_info.

Signed-off-by: vishaldhingra vdhingra@vmware.com

### CODE REVIEW

**Comment by vishaldhingra** [comment]
Testing 
Manual Testing 
1. IBGP with and without next hop self.
2. EBGP session 
3. Testing with 2K routes to simulate the work queue scenarios
4. Peer interface up/down
5. clear bgp session *
6. no neighbor
7. no bgp

File: bgpd/bgp_nexthop.h [hunk 1]
```diff
@@ -54,12 +54,16 @@ struct bgp_nexthop_cache {
...
 #define BGP_NEXTHOP_LABELED_VALID     (1 << 6)
+#define BGP_PEER_ADDED                (1 << 7)
+
 
 	uint16_t change_flags;
 
 #define BGP_NEXTHOP_CHANGED           (1 << 0)
 #define BGP_NEXTHOP_METRIC_CHANGED    (1 << 1)
 #define BGP_NEXTHOP_CONNECTED_CHANGED (1 << 2)
+#define BGP_PEER_DELETED              (1 << 2)
```
Comment (srimohans) [inline bgpd/bgp_nexthop.h]: Shouldn't this 1 << 3 ?

Comment (srimohans) [inline bgpd/bgp_nexthop.h]: Shouldn't this 1 << 3 ?

File: bgpd/bgp_nht.c [hunk 2]
```diff
@@ -88,7 +88,10 @@ int bgp_find_nexthop(struct bgp_info *path, int connected)
 
 static void bgp_unlink_nexthop_check(struct bgp_nexthop_cache *bnc)
 {
-	if (LIST_EMPTY(&(bnc->paths)) && bnc->nht_info) {
+	if (LIST_EMPTY(&(bnc->paths))
+		    && ((CHECK_FLAG(bnc->flags, BGP_PEER_ADDED)
```
Comment (srimohans) [inline bgpd/bgp_nht.c]: May be I am not completely following, but why will both ADDED & DELETED flags will be set in any scenario ? isn't this wrong ?

File: bgpd/bgp_nht.c [hunk 3]
```diff
@@ -134,6 +137,7 @@ void bgp_unlink_nexthop_by_peer(struct peer *peer)
 
 	/* cleanup the peer reference */
 	bnc->nht_info = NULL;
+	SET_FLAG(bnc->change_flags, BGP_PEER_DELETED);
```
Comment (srimohans) [inline bgpd/bgp_nht.c]: Should we clear the BGP_PEER_ADDED flag here  in addition to setting it to BGP_PEER_DELETED flag ?

File: bgpd/bgp_nht.c [hunk 4]
```diff
@@ -264,8 +268,10 @@ int bgp_find_or_add_nexthop(struct bgp *bgp_route, struct bgp *bgp_nexthop,
 			(bgp_info_extra_get(ri))->igpmetric = bnc->metric;
 		else if (ri->extra)
 			ri->extra->igpmetric = 0;
-	} else if (peer)
+	} else if (peer) {
 		bnc->nht_info = (void *)peer; /* NHT peer reference */
+		SET_FLAG(bnc->flags, BGP_PEER_ADDED);
```
Comment (srimohans) [inline bgpd/bgp_nht.c]: Same like above comment, Should we clear the BGP_PEER_DELETED flag here in addition to setting it to BGP_PEER_ADDED flag ?

File: bgpd/bgp_nht.c (hunk 2 above)
Comment (kssoman) [reply bgpd/bgp_nht.c]: The ADDED flag can be replaced with name BGP_NEXTHOP_PEER (to indicate that it is peer nexthop)  and deleted flag can be named as BGP_NEXTHOP_PEER_DELETED

When the peer gets established, the peer nexthop is registered to RIB and the ADDED flag will be set and DELETED flag is set when peer is being deleted and the  nexthop is unregistered from RIB

File: bgpd/bgp_nexthop.h (hunk 1 above)
Comment (vishaldhingra) [reply bgpd/bgp_nexthop.h]: yes, it would be 3 only, will update. Merging mistake sry

File: bgpd/bgp_nht.c (hunk 2 above)
Comment (srimohans) [reply bgpd/bgp_nht.c]: My question is different. Why they both get set at the **same time** ? The above has an AND check.

**Pull request closed without merging**